compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -503,6 +505,23 @@ fn layout_of_uncached<'tcx>(
                 ));
             }
 
+            let is_unsized_field = |field: &FieldDef| {
+                let param_env = tcx.param_env(def.did());
+                !tcx.type_of(field.did).instantiate_identity().is_sized(tcx, param_env)
```

Comment:
The issue with projections is that they must implement `Sized` on their own, meaning they must not have a `?Sized` if generic, or normalize to something that is known to be `Sized`.

So... basically what we'll have to do is invoke the query machinery and let it solve this for us:

```rust
tcx.infer_ctxt().build().type_implements_trait(sized_trait, [field_type], param_env).must_apply_modulo_regions()
```

side-note: you're right in using the struct's `param_env` as well as `instantiate_identity`, we need to only know this of the generic trait, not after instantiating its generic params.

Reply:
Thanks. Will try it out.

Reply:
I've changed the implementation of `is_unsized_field` to this:
```rust
            let is_unsized_field = |field: &FieldDef| {
                let param_env = tcx.param_env(def.did());
                let field_ty = tcx.type_of(field.did).skip_binder();
                let sized_trait = tcx.require_lang_item(hir::LangItem::Sized, None);
                !tcx.infer_ctxt().build().type_implements_trait(sized_trait, [field_ty], param_env).must_apply_modulo_regions()
            };
```
All the ICEs while building stdlib are gone but when I tried to build the below code `is_unsized_field` still returns `true` (i.e. `trait_implements_trait(...).must_apply_modulo_regions()` returns `false`) for the `StructWithProjections` struct below and it triggers may span delayed bug causing the ICE:

```rust
use std::mem::size_of;

struct StructWithProjections<'a>
{
    field1: i32,
    field2: <&'a [i32] as IntoIterator>::IntoIter,
    pub field3: [u8; 16],
}

pub fn main() {
    assert_eq!(size_of::<StructWithProjections>(), 21);
}
```

I couldn't figure out what I'm doing wrong.

Reply:
My next step would be to set `RUSTC_LOG=[evaluate_obligation]` (only for the command line the failing test is showing, and remember to dump to a file). This is a lot of info, and hard to process, but it should have all the info needed to see what's wrong. If you want we can do a screen share pairing session so you don't have to sit infront of this wall of text

Reply:
Thanks Oli. I'll try that out on my own first and contact you if I get stuck.

Reply:
I analyzed the logs of the compilation of the above snippet of code containing `StructWithProjections` having `field2` as a projection of type `<&'a [i32] as IntoIterator>::IntoIter`. The root cause of the failure seems to be a `Sorts` error while trying to relate `<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>`  as the super type of  `<std::slice::Iter<'_, i32> as std::marker::Sized>` (wherein the latter was obtained by normalizing the former). 

Here's the relevant subset of the logs:
```text
[compiler\rustc_ty_utils\src\layout.rs:514:17] ">>> is_unsized_field" = ">>> is_unsized_field"
[compiler\rustc_ty_utils\src\layout.rs:514:17] field.name = "field2"
│ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::infer::type_implements_trait trait_def_id=DefId(2:31977 ~ core[f031]::marker::Sized), param_env=ParamEnv { caller_bounds: [], reveal: UserFacing }
│ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::region type_flags(ReBound(DebruijnIndex(1), BoundRegion { var: 0, kind: BrAnon })) = TypeFlags(HAS_RE_BOUND)
│ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_traits::evaluate_obligation evaluate_obligation(canonical_goal=Canonical {
│ │ │ │ │ │ │ │ │ │     value: ParamEnvAnd {
│ │ │ │ │ │ │ │ │ │         param_env: ParamEnv {
│ │ │ │ │ │ │ │ │ │             caller_bounds: [],
│ │ │ │ │ │ │ │ │ │             reveal: UserFacing,
│ │ │ │ │ │ │ │ │ │         },
│ │ │ │ │ │ │ │ │ │         value: Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] },
│ │ │ │ │ │ │ │ │ │     },
│ │ │ │ │ │ │ │ │ │     max_universe: U0,
│ │ │ │ │ │ │ │ │ │     variables: [
│ │ │ │ │ │ │ │ │ │         CanonicalVarInfo {
│ │ │ │ │ │ │ │ │ │             kind: Region(
│ │ │ │ │ │ │ │ │ │                 U0,
│ │ │ │ │ │ │ │ │ │             ),
│ │ │ │ │ │ │ │ │ │         },
│ │ │ │ │ │ │ │ │ │     ],
│ │ │ │ │ │ │ │ │ │ })
│ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints created new region variable '?0 in U0 with origin MiscVariable(no-location (#0))
│ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::region type_flags('?0) = TypeFlags(HAS_RE_INFER | HAS_FREE_LOCAL_REGIONS | HAS_FREE_REGIONS)
│ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_traits::evaluate_obligation evaluate_obligation: goal=ParamEnvAnd {
│ │ │ │ │ │ │ │ │ │     param_env: ParamEnv {
│ │ │ │ │ │ │ │ │ │         caller_bounds: [],
│ │ │ │ │ │ │ │ │ │         reveal: UserFacing,
│ │ │ │ │ │ │ │ │ │     },
│ │ │ │ │ │ │ │ │ │     value: Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] },
│ │ │ │ │ │ │ │ │ │ }
│ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select with_query_mode, query_mode=Canonical
│ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::probe 
│ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer start_snapshot()
│ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: start_snapshot
│ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::evaluate_predicate_recursively obligation=Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0), previous_stack=None
│ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::evaluate_trait_predicate_recursively obligation=Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)
│ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select fresh_trait_pred=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select get_provisional = None, fresh_trait_pred=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::candidate_from_obligation stack=TraitObligationStack(Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0))
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select cache_fresh_trait_pred=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select is_knowable(intercrate=false)
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::candidate_assembly::assemble_candidates 
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:42:9] &obligation = Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:66:13] ">>> positive polarity" = ">>> positive polarity"
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::candidate_assembly::assemble_candidates_for_trait_alias 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select::candidate_assembly self_ty=Alias(Projection, AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) })
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:72:13] &obligation.predicate = Binder {
    value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive),
    bound_vars: [],
}
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:95:17] ">>> sized_trait" = ">>> sized_trait"
[compiler\rustc_trait_selection\src\traits\select\mod.rs:2104:9] &self_ty.kind() = Alias(
    Projection,
    AliasTy {
        args: [
        &'?0 [i32],
        ],
        def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter),
    },
)
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:99:17] &sized_conditions = None
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::candidate_assembly::assemble_builtin_bound_candidates conditions=None
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::candidate_assembly::assemble_candidates_from_projected_tys obligation=Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::probe 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer start_snapshot()
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: start_snapshot
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::higher_ranked::enter_forall_and_leak_universe binder=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::higher_ranked return=TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select::candidate_assembly placeholder_trait_predicate=TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::traits::util super_predicates, data=TraitPredicate(<<Self as std::iter::IntoIterator>::IntoIter as std::iter::Iterator>, polarity:Positive), obligations=Map { iter: Enumerate { iter: Iter([]), count: 0 } }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::traits::util super_predicates, data=TraitPredicate(<<Self as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), obligations=Map { iter: Enumerate { iter: Iter([]), count: 0 } }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::generic_args shift_vars(val=&'?0 [i32], binders_passed=1, has_escaping_bound_vars=false)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::generic_args shift_vars(val=&'?0 [i32], binders_passed=1, has_escaping_bound_vars=false)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::generic_args shift_vars(val=&'?0 [i32], binders_passed=1, has_escaping_bound_vars=false)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::generic_args shift_vars(val=&'?0 [i32], binders_passed=1, has_escaping_bound_vars=false)
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:199:21] &bound = Binder { value: ProjectionPredicate(AliasTy { args: [Alias(Projection, AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) })], def_id: DefId(2:8256 ~ core[f031]::iter::traits::iterator::Iterator::Item) }, Term::Ty(Alias(Projection, AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8109 ~ core[f031]::iter::traits::collect::IntoIterator::Item) }))), bound_vars: [] }
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:201:25] ">>> as_trait_clause()" = ">>> as_trait_clause()"
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:199:21] &bound = Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::iter::Iterator>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::probe 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer start_snapshot()
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: start_snapshot
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::match_normalize_trait_ref obligation=Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0), trait_bound=Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::iter::Iterator>, bound_vars: [] }, placeholder_trait_ref=<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1661:9] ">> match_normalize_trait_ref" = ">> match_normalize_trait_ref"
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1661:9] &obligation = Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1661:9] &trait_bound = Binder {
    value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::iter::Iterator>,
    bound_vars: [],
}
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1661:9] &placeholder_trait_ref = <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1664:13] ">> match_normalize_trait_ref returning early" = ">> match_normalize_trait_ref returning early"
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select return=Err(())
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::rollback_to 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::undo_log rollback_to(0)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: rollback_to(RegionSnapshot)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
[compiler\rustc_trait_selection\src\traits\select\candidate_assembly.rs:199:21] &bound = Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::probe 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer start_snapshot()
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: start_snapshot
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::match_normalize_trait_ref obligation=Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0), trait_bound=Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }, placeholder_trait_ref=<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1661:9] ">> match_normalize_trait_ref" = ">> match_normalize_trait_ref"
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1661:9] &obligation = Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1661:9] &trait_bound = Binder {
    value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>,
    bound_vars: [],
}
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1661:9] &placeholder_trait_ref = <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1670:13] ">>> normalize_with_depth" = ">>> normalize_with_depth"
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::normalize::normalize_with_depth_to depth=1, value=Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize obligations.len=0
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize value=Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::project::opt_normalize_projection_type projection_ty=AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) }, depth=1
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::project no cache
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::project::project obligation=Obligation(predicate=AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) }, depth=1)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::project::assemble_candidates_from_predicates obligation=Obligation(predicate=AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) }, depth=1)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::project assemble_candidates_from_trait_def(..)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::project assemble_candidates_from_object_ty(..)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::project::assemble_candidates_from_impls 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::commit_if_ok 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer start_snapshot()
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: start_snapshot
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::poly_select obligation=Obligation(predicate=Binder { value: TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive), bound_vars: [] }, depth=1)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::candidate_from_obligation stack=TraitObligationStack(Obligation(predicate=Binder { value: TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive), bound_vars: [] }, depth=1))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select cache_fresh_trait_pred=Binder { value: TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select CACHE HIT
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select return=Ok(Some(ImplCandidate(DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}))))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::confirmation::confirm_candidate obligation=Obligation(predicate=Binder { value: TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive), bound_vars: [] }, depth=1), candidate=ImplCandidate(DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select::confirmation confirm_impl_candidate, obligation=Obligation(predicate=Binder { value: TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive), bound_vars: [] }, depth=1), impl_def_id=DefId(2:10987 ~ core[f031]::slice::iter::{impl#0})
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::match_impl impl_def_id=DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}), impl_trait_header=EarlyBinder { value: ImplTraitHeader { trait_ref: <&'a [T] as std::iter::IntoIterator>, polarity: Positive, unsafety: Normal } }, obligation=Obligation(predicate=Binder { value: TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive), bound_vars: [] }, depth=1)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::higher_ranked::enter_forall_and_leak_universe binder=Binder { value: TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::higher_ranked return=TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints created new region variable '?1 in U0 with origin RegionParameterDefinition(no-location (#0), "'a")
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::type_variable new_var(index=?0t, universe=U0, origin=TypeVariableOrigin { kind: TypeParameterDefinition("T", DefId(2:10989 ~ core[f031]::slice::iter::{impl#0}::T)), span: no-location (#0) })
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::generic_args shift_vars(val=?0t, binders_passed=0, has_escaping_bound_vars=false)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::region type_flags('?1) = TypeFlags(HAS_RE_INFER | HAS_FREE_LOCAL_REGIONS | HAS_FREE_REGIONS)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select impl_trait_header=ImplTraitHeader { trait_ref: <&[_] as std::iter::IntoIterator>, polarity: Positive, unsafety: Normal }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::normalize::normalize_with_depth_to depth=2, value=<&[_] as std::iter::IntoIterator>
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize obligations.len=0
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize value=<&[_] as std::iter::IntoIterator>
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize result=<&[_] as std::iter::IntoIterator>, obligations.len=0
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize normalizer.obligations=[]
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select impl_trait_ref=<&[_] as std::iter::IntoIterator>, placeholder_obligation_trait_ref=<&[i32] as std::iter::IntoIterator>
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::at::eq define_opaque_types=No, a=<&[i32] as std::iter::IntoIterator>, b=<&[_] as std::iter::IntoIterator>
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::equate::tys a=&'?0 [i32], b=&'?1 [?0t]
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::equate Equate.regions('?0, '?1)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::region_constraints::make_subregion sub='?0, sup='?1
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints origin = Subtype(
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │     TypeTrace {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │         cause: ObligationCause {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             span: no-location (#0),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             body_id: DefId(0:0 ~ ice_non_last_unsized_field_issue_121473[4941]),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             code: MatchImpl(
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 ObligationCause {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     span: no-location (#0),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     body_id: DefId(0:0 ~ ice_non_last_unsized_field_issue_121473[4941]),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     code: MiscObligation,
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             ),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │         },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │         values: PolyTraitRefs(
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             ExpectedFound {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 expected: Binder {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     value: <&[i32] as std::iter::IntoIterator>,
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     bound_vars: [],
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 found: Binder {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     value: <&[_] as std::iter::IntoIterator>,
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     bound_vars: [],
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │         ),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │     },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ )
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: add_constraint(VarSubVar('?0, '?1))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::region_constraints::make_subregion sub='?1, sup='?0
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints origin = Subtype(
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │     TypeTrace {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │         cause: ObligationCause {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             span: no-location (#0),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             body_id: DefId(0:0 ~ ice_non_last_unsized_field_issue_121473[4941]),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             code: MatchImpl(
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 ObligationCause {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     span: no-location (#0),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     body_id: DefId(0:0 ~ ice_non_last_unsized_field_issue_121473[4941]),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     code: MiscObligation,
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             ),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │         },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │         values: PolyTraitRefs(
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             ExpectedFound {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 expected: Binder {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     value: <&[i32] as std::iter::IntoIterator>,
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     bound_vars: [],
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 found: Binder {
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     value: <&[_] as std::iter::IntoIterator>,
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                     bound_vars: [],
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │                 },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │             },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │         ),
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │     },
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ )
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: add_constraint(VarSubVar('?1, '?0))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints make_eqregion: unifying '?0 with '?1
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::relate Equate.mts(TypeAndMut { ty: [i32], mutbl: Not }, TypeAndMut { ty: [?0t], mutbl: Not })
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::equate::tys a=[i32], b=[?0t]
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::equate::tys a=i32, b=?0t
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::generalize::instantiate_ty_var target_is_expected=false, target_vid=?0t, instantiation_variance=o, source_ty=i32
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::generalize::tys t=i32
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::generalize return=Ok(i32)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::generalize flip relation
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::equate::tys a=i32, b=i32
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select return=Ok(Normalized(['?1, ?0t], []))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select::confirmation impl args, args=Normalized(['?1, ?0t], [])
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select::confirmation vtable_impl, impl_def_id=DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}), args=Normalized(['?1, ?0t], []), recursion_depth=2
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::impl_or_trait_obligations recursion_depth=2, def_id=DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}), args=['?1, ?0t], parent_trait_pred=Binder { value: TraitPredicate(<&[i32] as std::iter::IntoIterator>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::generic_args shift_vars(val=?0t, binders_passed=1, has_escaping_bound_vars=false)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::normalize::normalize_with_depth_to depth=2, value=Binder { value: TraitPredicate(<_ as std::marker::Sized>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize obligations.len=0
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize value=Binder { value: TraitPredicate(<i32 as std::marker::Sized>, polarity:Positive), bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize result=Binder { value: TraitPredicate(<i32 as std::marker::Sized>, polarity:Positive), bound_vars: [] }, obligations.len=0
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize normalizer.obligations=[]
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select::confirmation vtable_impl, impl_obligations=[Obligation(predicate=Binder { value: TraitPredicate(<i32 as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=2)]
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select return=Ok(Some(ImplSourceUserDefinedData(impl_def_id=DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}), args=['?1, ?0t], nested=[Obligation(predicate=Binder { value: TraitPredicate(<i32 as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=2)])))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer commit_if_ok() -- r.is_ok() = true
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::commit_from 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::undo_log commit(1)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::project confirm_candidate, obligation=Obligation(predicate=AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) }, depth=1), candidate=Select(ImplSourceUserDefinedData(impl_def_id=DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}), args=['?1, ?0t], nested=[Obligation(predicate=Binder { value: TraitPredicate(<i32 as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=2)]))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::specialize translate_args(ParamEnv { caller_bounds: [], reveal: UserFacing }, DefId(2:10987 ~ core[f031]::slice::iter::{impl#0}), ['?1, ?0t], Impl(DefId(2:10987 ~ core[f031]::slice::iter::{impl#0})))
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::generic_args shift_vars(val=?0t, binders_passed=0, has_escaping_bound_vars=false)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::generic_args shift_vars(val=?0t, binders_passed=0, has_escaping_bound_vars=false)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::region type_flags('?1) = TypeFlags(HAS_RE_INFER | HAS_FREE_LOCAL_REGIONS | HAS_FREE_REGIONS)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_middle::ty::region type_flags('?1) = TypeFlags(HAS_RE_INFER | HAS_FREE_LOCAL_REGIONS | HAS_FREE_REGIONS)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::traits::project ProjectionCacheEntry::insert_ty: adding cache entry: key=ProjectionCacheKey { ty: AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) } }, value=Normalized(Term::Ty(std::slice::Iter<'?1, i32>), [Obligation(predicate=Binder { value: TraitPredicate(<i32 as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=2)])
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize AssocTypeNormalizer: normalized type, self.depth=1, ty=Alias(Projection, AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) }), normalized_ty=Term::Ty(std::slice::Iter<'?1, i32>), obligations.len=1
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize result=Binder { value: <std::slice::Iter<'_, i32> as std::marker::Sized>, bound_vars: [] }, obligations.len=1
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::normalize normalizer.obligations=[Obligation(predicate=Binder { value: TraitPredicate(<i32 as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=2)]
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1679:9] &trait_bound = Binder {
    value: <std::slice::Iter<'_, i32> as std::marker::Sized>,
    bound_vars: [],
}
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::at::sub define_opaque_types=No, a=Binder { value: <std::slice::Iter<'_, i32> as std::marker::Sized>, bound_vars: [] }, b=Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::higher_ranked::higher_ranked_sub sub=Binder { value: <std::slice::Iter<'_, i32> as std::marker::Sized>, bound_vars: [] }, sup=Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }, sub_is_expected=false
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::higher_ranked::enter_forall forall=Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::higher_ranked::enter_forall_and_leak_universe binder=Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::higher_ranked return=<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::higher_ranked ?value
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::higher_ranked a_prime=<std::slice::Iter<'_, i32> as std::marker::Sized>
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::relate::higher_ranked b_prime=<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::relate::equate::tys a=std::slice::Iter<'?1, i32>, b=Alias(Projection, AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) })
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1695:17] ">>> sup relate failed" = ">>> sup relate failed"
[compiler\rustc_trait_selection\src\traits\select\mod.rs:1695:17] &e = Sorts(
    ExpectedFound {
        expected: Alias(
            Projection,
            AliasTy {
                args: [
                &'?0 [i32],
                ],
                def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter),
            },
        ),
        found: std::slice::Iter<'?1, i32>,
    },
)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select return=Err(())
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::rollback_to 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::undo_log rollback_to(0)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: rollback_to(RegionSnapshot)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::rollback_to 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::undo_log rollback_to(0)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: rollback_to(RegionSnapshot)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::candidate_assembly::assemble_candidates_from_caller_bounds 
│ │ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select::candidate_assembly stack.obligation=Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select::candidate_assembly assemble_candidates_from_auto_impls, self_ty=Alias(Projection, AliasTy { args: [&'?0 [i32]], def_id: DefId(2:8110 ~ core[f031]::iter::traits::collect::IntoIterator::IntoIter) })
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select::candidate_assembly candidate list size: 0
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select assembled 0 candidates, stack=TraitObligationStack(Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)), candidates=[]
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::filter_impls obligation=Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select winnowed to 0 candidates, stack=TraitObligationStack(Obligation(predicate=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, depth=0)), candidates=[]
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select CACHE MISS
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─┐rustc_trait_selection::traits::select::insert_candidate_cache candidate=Err(Unimplemented)
│ │ │ │ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_trait_selection::traits::select insert_candidate_cache global, pred=TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), candidate=Err(Unimplemented)
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select return=Err(Unimplemented)
│ │ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select on_failure, dfn=0
│ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select CACHE MISS
│ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select insert_evaluation_cache global, trait_pred=Binder { value: TraitPredicate(<<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, polarity:Positive), bound_vars: [] }, result=EvaluatedToErr
│ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select on_completion, dfn=0
│ │ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select return=Ok(EvaluatedToErr)
│ │ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::traits::select return=Ok(EvaluatedToErr)
│ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::region_constraints::leak_check::leak_check outer_universe=U0, max_universe=U0
│ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints::leak_check return=Ok(())
│ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ │ ├─┐rustc_infer::infer::rollback_to 
│ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::undo_log rollback_to(0)
│ │ │ │ │ │ │ │ │ │ │ ├─  0ms DEBUG rustc_infer::infer::region_constraints RegionConstraintCollector: rollback_to(RegionSnapshot)
│ │ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ ├─┘
│ │ │ │ │ │ │ │ │ ├─  1ms DEBUG rustc_trait_selection::infer return=EvaluatedToErr
│ │ │ │ │ │ │ │ ├─┘

```

Reply:
The summary of the above trace is that while trying to relate `Binder { value: <std::slice::Iter<'_, i32> as std::marker::Sized>, bound_vars: [] }` to `Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }` as a subtype infcx goes through various motions and ultimately tries to relate them structurally in which it fails which causes the evaluate machinery to deduce that the projection type is unsized.

This seems to be a pre-existing issue though. In fact the `is_sized()` call in the existing code in `layout.rs` shown below https://github.com/rust-lang/rust/blob/dc2ffa405407ffb3654658c50ab3dfda124fbdfd/compiler/rustc_ty_utils/src/layout.rs#L524-L528
uses the exact same evaluate logic as the above and fails precisely the same way. That is, `is_sized()` too returns `false` when called on a projection type like `<&'a [i32] as IntoIterator>::IntoIter`. This does not lead to a catastrophic failure however. It merely causes a slightly different layout to be chosen as the below logic over in 'rustc_abi` gets skipped: https://github.com/rust-lang/rust/blob/dc2ffa405407ffb3654658c50ab3dfda124fbdfd/compiler/rustc_abi/src/layout.rs#L97
However, in our case it causes a delay bug to be triggered.

@oli-obk In your opinion is this failure to relate `Binder { value: <std::slice::Iter<'_, i32> as std::marker::Sized>, bound_vars: [] }`  to `Binder { value: <<&[i32] as std::iter::IntoIterator>::IntoIter as std::marker::Sized>, bound_vars: [] }` as a subtype a bug or expected behaviour?

Reply:
kind of. I expect these to be different, I just don't expect the latter to actually occur, and instead expected it to get normalized. I mean, something got normalized somewhere as we got the `slice::Iter`.

Note that we do already normalize at the start of the `layout_of` query, so maybe what we actually need to do is to invoke `layout_of` on the instantiated type and ask the layout whether it is sized?

Reply:
I tried `layout_of(field_ty).is_sized()` as shown below:

```rust
            let is_unsized_field = |field: &FieldDef| {
                let param_env = tcx.param_env(def.did());
                let field_ty = tcx.type_of(field.did).instantiate_identity();
                tcx.layout_of(ParamEnvAnd { param_env, value: field_ty })
                    .map(|field_layout| !field_layout.is_sized())
                    .or(Err(error(cx, LayoutError::Unknown(ty))))
            };
```

but it unfortunately also causes loads of ICEs.

Wonder if we have to change the approach and try to taint the typeck results instead. Perhaps that is a general strategy we should adopt to keep malformed types from reaching the const eval/mir building stage as #121097 might also be helped by something like that.

Reply:
> Wonder if we have to change the approach and try to taint the typeck results instead

Reply:
If we let wfcheck set a flag on, say `hir::Ty`, that the type is not WF then typeck can use that to taint its results. 

For example, let's say we add a `wf_check_failed` flag to `hir::Ty`:

```rust
pub struct Ty<'hir> {
    pub hir_id: HirId,
    pub kind: TyKind<'hir>,
    pub span: Span,
    pub bool: wf_check_failed,
}
```

And then in wfcheck set this flag on failure:

```rust
if /* wf check failed for ty */ {
    ty.wf_check_failed = true
}
```

Then somewhere in typeck check the flag and taint the results:
```rust
if ty.wf_check_failed && /* some other appropriate condition */ {
   // taint typeck results
}
```
Something like that can provide a general way to fix MIR ICEs caused early errors.

Reply:
that's mutation of global state which violates all kinds of incremental/query-caching rules we have. Even if technically fine due to we already being in an erroring compilation and will never write it to the cache anyway.

the main reason we shouldn't do it is that the result of typeck would then differ depending on whether wfcheck had already been run.

Reply:
Ah, make sense.